crates/shim-protos/build.rs

```diff
@@ -37,42 +48,126 @@ fn main() {
             "vendor/github.com/containerd/containerd/api/events/task.proto",
         ],
+        true,
         false,
     );
 
     codegen(
         "src/cgroups",
         &["vendor/github.com/containerd/cgroups/stats/v1/metrics.proto"],
         true,
+        false,
     );
 
+    // generate async service
     codegen(
-        "src/shim",
+        "src/services",
+        &[
+            "vendor/github.com/containerd/containerd/runtime/v2/task/shim.proto",
+            "vendor/github.com/containerd/containerd/api/services/ttrpc/events/v1/events.proto",
+        ],
+        true,
+        true,
+    );
+    //  rename to async
+    fs::rename(
+        "src/services/shim_ttrpc.rs",
+        "src/services/shim_ttrpc_async.rs",
+    )
+    .unwrap();
+    fs::rename(
+        "src/services/events_ttrpc.rs",
+        "src/services/events_ttrpc_async.rs",
+    )
+    .unwrap();
+
+    // generate sync service
+    codegen(
+        "src/services",
         &[
             "vendor/github.com/containerd/containerd/runtime/v2/task/shim.proto",
-            "vendor/github.com/containerd/containerd/api/types/mount.proto",
-            "vendor/github.com/containerd/containerd/api/types/task/task.proto",
             "vendor/github.com/containerd/containerd/api/services/ttrpc/events/v1/events.proto",
-            "vendor/google/protobuf/empty.proto",
         ],
         true,
+        false,
     );
 
+    // replace text in services/shim.rs
+    replace_text_in_file(
+        "src/services/shim.rs",
+        "super::mount",
+        "crate::types::mount",
+    )
+    .unwrap();
+    replace_text_in_file("src/services/shim.rs", "super::task", "crate::types::task").unwrap();
+    replace_text_in_file(
+        "src/services/shim.rs",
+        "super::empty",
+        "crate::types::empty",
+    )
+    .unwrap();
+
+    // replace text in shim/shim_ttrpc.rs
+    replace_text_in_file(
+        "src/services/shim_ttrpc.rs",
+        "super::empty",
+        "crate::types::empty",
+    )
+    .unwrap();
+
+    // replace text in shim/shim_ttrpc_async.rs
+    replace_text_in_file(
+        "src/services/shim_ttrpc_async.rs",
+        "super::empty",
+        "crate::types::empty",
+    )
+    .unwrap();
+
+    // replace text in shim/events_ttrpc.rs
+    replace_text_in_file(
+        "src/services/events_ttrpc.rs",
+        "super::empty",
+        "crate::types::empty",
+    )
+    .unwrap();
+
+    // replace text in shim/events_ttrpc_async.rs
+    replace_text_in_file(
+        "src/services/events_ttrpc_async.rs",
+        "super::empty",
+        "crate::types::empty",
+    )
+    .unwrap();
+
+    // replace text in shim/events/events/task.rs
+    replace_text_in_file("src/events/task.rs", "super::mount", "crate::types::mount").unwrap();
+
     // TODO: shim_ttrpc is not included in mod.rs, file a bug upstream.
     let mut f = fs::OpenOptions::new()
         .append(true)
-        .open("src/shim/mod.rs")
+        .open("src/services/mod.rs")
```

Comment:
Will it get simpler by manually maintain the src/services/mod.rs instead of auto-generate?

Reply:
Currently, the auto-generate services (shim_ttrpc.rs, events_ttrpc.rs) are not included in mod.rs, so have to be manually maintained.